Competitive_Programming/C++/Searching & Sorting/Merge_Sort_Implementation_and_analysis.md

```diff
@@ -0,0 +1,118 @@
+## Merge Sort and Its Analysis
+### Overview
+Like QuickSort, Merge Sort is a Divide and Conquer algorithm. It divides the input array into two halves, calls itself for the two halves, and then merges the two sorted halves. The merge() function is used for merging two halves. The merge(arr, l, m, r) is a key process that assumes that arr[l..m] and arr[m+1..r] are sorted and merges the two sorted sub-arrays into one.
+
+### Algorithm
+MergeSort(arr[], l,  r) <br/>
+If r > l <br/>
+     1. Find the middle point to divide the array into two halves: <br/>  
+             middle m = l+ (r-l)/2 <br/>
+     2. Call mergeSort for first half:   <br/>
+             Call mergeSort(arr, l, m) <br/>
+     3. Call mergeSort for second half: <br/>
+             Call mergeSort(arr, m+1, r) <br/>
+     4. Merge the two halves sorted in step 2 and 3: <br/>
+             Call merge(arr, l, m, r) <br/>
+<hr>
+<p align="center">
+<img src="https://user-images.githubusercontent.com/63805002/134805767-7f0ec507-444e-45b6-9c06-5c1914c3d7c8.png"> 
+</p>
+
+
+### Code.
+```
+#include<iostream>
+#include<vector>
+
+using namespace std;
+
+//Helper Method
+void merge(vector<int> &array, int s, int e){
+    int i = s;
+    int m = (s+e)/2;
+    int j = m + 1;
+
+    vector<int> temp;
+    while(i<=m and j<=e){
+        if(array[i] < array[j]){
+            temp.push_back(array[i]);
+            i++;
+        }
+        else{
+            temp.push_back(array[j]);
+            j++;
+        }
+    }
+    //copy remaining element from first array
+    while(i<=m){
+        temp.push_back(array[i++]);
+    }
+
+    //copy remaining element from second array
+    while(j<=e){
+        temp.push_back(array[j++]);
+    }
+
+    //copy back the element from temp to original array
+    int k = 0;
+    for(int idx = s; idx <= e; ++idx){
+        array[idx] = temp[k];
+        k++;
+    }
+    return;
+}
+
+//sorting method
+void mergesort(vector<int> &arr, int s, int e){
+    //base case i.e if there is only one element in array so its already sorted
+    if(s>=e){
+        return;
+    }
+
+    //rec case
+    int mid = (s+e)/2;
+    mergesort(arr,s,mid);
+    mergesort(arr,mid+1,e);
+    return merge(arr,s,e);
+}   
+
+int main(){
+    vector<int> arr{10,5,2,0,7,6,4};
+
+    int s = 0;
+    int e = arr.size()-1;
+    mergesort(arr,s,e);
+    for(int i : arr){
+        cout<<i<<" ,";
+    }
+}
+```
+
```

Comment:
Add `input/output` analysis as well

Reply:
thanks @ravireddy07 for pointing out the mistakes. I have made the respective changes please see to it @ravireddy07